### DISCUSSION TITLE
Updates to Assessment Data Model (circa v1.9)

### DISCUSSION THREAD

**Discussion by jwildfire** [Gilead-Public/gsm#1506] [Ideas]
# Overview 

We're considering some updates that will further standardize the {gsm} assessment data model. The primary goal is to allow for a standardized process for drilling down see site- and  participant- level trends across all KRIs. 

# Background

The mapping process is currently not standard across KRIs. There are 2 basic approaches that can be shown by comparing the AE and Labs mappings. 

The AE mapping function takes 2 datasets (`AE_Map_Raw(dfs = list(dfSUBJ = "clindata::rawplus_dm", dfAE ="clindata::rawplus_ae")),...))`) whose rows are then aggregated to get numerator (AE count) and denominator (time on study) for the AE KRI. This approach allows us to look at trends in the numerator for sites/participants of interest by simply examining dfAE. 

The lab mapping function also takes two domains (`LB_Map_Raw(dfs = list(dfSUBJ = "clindata::rawplus_dm", dfLB = "clindata::rawplus_lb")))`), but `dfSUBJ` is only used to filter on enrollment status. Both the numerator (labs at grade 4+) and denominator (total labs) are generated using `dfLB` like so: 

```
    dfInput <- dfLB_mapped %>%
      mutate(
        Count = if_else(
          .data$Grade %in% lMapping[["dfLB"]][["strGradeHighVal"]],
          1,
          0
        ),
        Total = 1
      ) %>%
      group_by(.data$SubjectID) %>%
      summarize(
        Count = sum(.data$Count, na.rm = TRUE),
        Total = sum(.data$Total, na.rm = TRUE)
      ) %>%
      ungroup() %>%
      gsm::MergeSubjects(dfSUBJ_mapped, vFillZero = "Count", vRemoval = "Total", bQuiet = bQuiet) %>%
      select(any_of(c(names(dfSUBJ_mapped))), "Count", "Total") %>%
      arrange(.data$SubjectID)
```

This works well for generating the needed counts, but doesn't record which Lab values were flagged as Grade 4+ in the data model. This greatly reduces our ability to drill down to site- and participant- level details in for risk signals of interest. 

# Proposed Change - Mapped Data Object

The biggest proposed change is adding a new "Mapped Data" layer to our data model. The updated data model would be: 

- Raw Data
- Mapped Data **New**
   - lMapped$numerator - One record per numerator item
   - lMapped$denominator - One record per denominator item
- Input Data
- Transformed Data
- Analyzed Data
- Flagged Data
- Summary Data

Mapping functions would be refactored to return a new list of data frames that capture the numerator and denominator. For example, the defaults for AE and Labs would look like: 

```
# AEs
lMapped <- list(
dfNumerator =clindata::rawplus_ae, 
dfDenominoator = clindata::rawplus_dm
)

# Labs
lMapped <- list(
   dfNumerator =clindata::rawplus_lb %>% filter(.data$Grade %in% lMapping[["dfLB"]][["strGradeHighVal"]]),  
   dfDenominoator = clindata::rawplus_lb
)
```

Input data would then be derived using a standardized function across all KRIs*. Something like

```
Make_Input_Rate(dfSubj, dfNumerator, strNumeratorCol=NULL, dfDenominator, strNumeratorCol=NULL, mapping,...){

    # Get standard participant details
    dfSUBJ_mapped <- dfs$dfSUBJ %>%
      select(
        SubjectID = lMapping[["dfSUBJ"]][["strIDCol"]],
        any_of(
          c(
            SiteID = lMapping[["dfSUBJ"]][["strSiteCol"]],
            StudyID = lMapping[["dfSUBJ"]][["strStudyCol"]],
            CountryID = lMapping[["dfSUBJ"]][["strCountryCol"]],
            CustomGroupID = lMapping[["dfSUBJ"]][["strCustomGroupCol"]]
          )
        )
      )

  #Calculate Numerator

      dfInput <- dfNumerator %>%
         group_by(.data$SubjectID) %>%
         mutate(numerator=  if_else(isNull(strNumeratorCol), n(), sum(strNumeratorCol) %>% # count rows or sum specified column
         ungroup() %>%
         gsm::MergeSubjects(dfSUBJ_mapped, vFillZero = "Count", vRemoval = "Total", bQuiet = bQuiet)
 
 #Calculate Denominator
 # strDenominatorCol == study_days (for AEs)
 # strDenominatorCol == NULL for treatment discontinuation
      dfInput <- dfDenominator %>%
         group_by(.data$SubjectID) %>%
         mutate(denominator=  if_else(isNull(strDenominatorCol), n(), sum(strDenominatorCol)%>%
         ungroup() %>%
         gsm::MergeSubjects(dfSUBJ_mapped, vFillZero = "Count", vRemoval = "Total", bQuiet = bQuiet)

return(dfInput)
}
```

Notes: 
- The rest of the assessment pipeline would need only minor changes. 
- The assessment functions themselves could be mostly unchanged if we called `Make_Input_Rate` upstream.
- Other `Make_Input_XXX` functions, with different `lMapping` structures could be created if we ever have KRIs that aren't rates. 
- This would standardize the columns in dfInput across all KRIs (this would require some minor refactor in assess functions)
- Drill down reports could check for whether lMapping is present before providing certain functionality
- We should consider the best way to make sure desired covariates are specified/included as part of the refactored mapping functions.
- Users could ignore this new framework and just provide a dfInput to Assess using whatever method they like (just like they don't have to use our current mapping functions). I do think we would move towards using this as our standard at Gilead.  
- I think this new process could be implemented fairly easily by adding a few steps in our existing workflows ...

**Comment by jwildfire** [comment]
@mattroumaya @samussiah Let me know what you think of this. I'm leaning towards implementing, but recognize that it is a fairly major refactor that will require updates to qualification for mapping.

**Comment by mattroumaya** [comment]
@jwildfire - starting to test this out now and will keep updating this with some notes for both of our reference.

1. `SubjectID` mappings differ across domains:

Example: 

`AE_Map_Raw()`: strIDCol

```
SubjectID = lMapping[["dfSUBJ"]][["strIDCol"]]
```

`QueryAge_Map_Raw()`: strEDCIDCol

```
SubjectID = lMapping[["dfSUBJ"]][["strEDCIDCol"]]
```

**Comment by jwildfire** [comment]
Partial implementation in #1509
